src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -122,48 +137,348 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
+      auto compressed_block = pmr_vector<char>{values.get_allocator()};
+      compressed_block.resize(block_bound);
+
+      /**
+       * If we previously learned a dictionary, we use it to initialize LZ4. Otherwise LZ4 uses the previously
+       * compressed block instead, which would cause the blocks to depend on one another.
+       * If we have no dictionary present and compress at least a second block (i.e., block_index > 0), then we reset
+       * the LZ4 stream to maintain the independence of the blocks. This only happens when the column does not contain
+       * enough data to produce a zstd dictionary (i.e., a column of single character strings).
+       */
+      if (!dictionary.empty()) {
+        LZ4_loadDictHC(lz4_stream, dictionary.data(), static_cast<int>(dictionary.size()));
+      } else if (block_index) {
+        LZ4_resetStreamHC(lz4_stream, LZ4HC_CLEVEL_MAX);
+      }
+
+      // The offset in the source data where the current block starts.
+      const auto value_offset = block_index * _block_size;
+      // move pointer to start position and pass to the actual compression method
+      const int compression_result = LZ4_compress_HC_continue(
+          lz4_stream, reinterpret_cast<char*>(values.data()) + value_offset, compressed_block.data(),
+          static_cast<int>(decompressed_block_size), static_cast<int>(block_bound));
+
+      Assert(compression_result > 0, "LZ4 stream compression failed");
+
+      // shrink the block vector to the actual size of the compressed result
+      compressed_block.resize(static_cast<size_t>(compression_result));
+      compressed_block.shrink_to_fit();
+
+      lz4_blocks.emplace_back(std::move(compressed_block));
+    }
+
+    // Finally, release the LZ4 stream memory.
+    LZ4_freeStreamHC(lz4_stream);
+  }
+
+  /**
+   * Generate a dictionary for non-strings. The zstd dictionary is intended for string data. Therefore, the samples
+   * provided to the zstd algorithm are not the values of each row but multiple values at once (since the minimum size
+   * for a sample is 8 bytes).
+   *
+   * @tparam T The data type of the value segment. This method is only called for non-string-segments.
+   * @param values All values of the segment. They are the input data to train the dictionary.
+   * @return The trained dictionary, or in the case of failure, an empty vector.
+   */
+  template <typename T>
+  pmr_vector<char> _train_dictionary(const pmr_vector<T>& values) {
+    /**
+     * The minimum sample size is 8 bytes. Since non-string data has a constant size for each value we can just set
+     * the sample size to 8 for all values.
+     */
+    const size_t min_sample_size = 8u;
+    const auto values_size = values.size() * sizeof(T);
+    const auto sample_size = std::max(sizeof(T), min_sample_size);
+    const auto num_samples = values_size / sample_size;
+    const std::vector<size_t> sample_lens(num_samples, sample_size);
```

Comment:
Wha is `sample_lens` supposed to mean?

Further, for a `long` column, we gather the full segment as samples (100k) and for `int`, we gather 50k samples. Why that?

Reply:
We renamed it to `sample_sizes`. It is the length of each sample. I.e., the value at index 0 is the size of the first sample in bytes. The samples are stored in the `values` vector.

Since the minimum sample size is 8 byte and a (32 bit) integer is only 4 bytes large, each sample consists of 2 integers. Thus we only have half as many samples.